compiler/optimizations/optimizeOnClauses.cpp

```diff
@@ -31,25 +31,41 @@
 #include "stmt.h"
 #include "passes.h"
 
-/* Print lots of debugging messages when DEBUG is defined */
-#ifdef DEBUG
-  #define DEBUG_PRINTF(...) printf(__VA_ARGS__)
-#else
-  #define DEBUG_PRINTF(...)
-#endif
+// There are two questions:
+// 1- is the primitive/function eligible to run in a fast block?
+//    any blocking or system call disqualifies it since a fast
+//    AM handler can be run in a signal handler
+// 2- is the primitive/function communication free?
+
+// Any function containing comminication can't run in a fast block.
+
+enum {
+  // The primitive is ineligible for a fast (e.g. uses a lock or allocator)
+  // AND it causes communication
+  NOT_FAST_NOT_LOCAL,
+  // Is the primitive ineligible for a fast (e.g. uses a lock or allocator)
+  // but communication free?
+  LOCAL_NOT_FAST,
+  // Does the primitive communicate?
+  // This implies NOT_FAST, unless it is in a local block
+  // if it is in a local block, this means IS_FAST.
+  FAST_NOT_LOCAL,
+  // Is the primitive function fast (ie, could it be run in a signal handler)
+  // IS_FAST implies IS_LOCAL.
+  FAST_AND_LOCAL
+};
 
 //
-// Return true if this primitive function is safe for fast on optimization
-// (e.g., no communication, no sync/single accesses)
+// Return NOT_FAST, NOT_LOCAL, IS_LOCAL, or IS_FAST.
 //
-static bool
-isFastPrimitive(CallExpr *call, bool isLocal) {
+static int
+classifyPrimitive(CallExpr *call) {
   INT_ASSERT(call->primitive);
   // Check primitives for communication
   switch (call->primitive->tag) {
   case PRIM_UNKNOWN:
     // TODO: Return true for PRIM_UNKNOWNs that are side-effect free
```

Comment:
Actually it's current. PRIM_UNKNOWN in some confusing way is a way to call some runtime functions. There are a bunch of runtime functions that get registered as primitives but don't have their own PRIM_... enum element.